terraform/transform_targets.go

```diff
@@ -133,25 +127,27 @@ func (t *TargetsTransformer) nodeIsTarget(v dag.Vertex, targets []addrs.Targetab
 		vertexAddr = r.ResourceInstanceAddr()
 	case GraphNodeConfigResource:
 		vertexAddr = r.ResourceAddr()
+
 	default:
 		// Only resource and resource instance nodes can be targeted.
 		return false
 	}
 
 	for _, targetAddr := range targets {
-		if t.IgnoreIndices {
-			// If we're ignoring indices then we'll convert any resource instance
-			// addresses into resource addresses. We don't need to convert
-			// vertexAddr because instance addresses are contained within
-			// their associated resources, and so .TargetContains will take
-			// care of this for us.
-			switch instance := targetAddr.(type) {
+		switch vertexAddr.(type) {
+		case addrs.ConfigResource:
+			// Before expansion happens, we only have nodes that know their
+			// ConfigResource address.  We need to take the more specific
+			// target addresses and generalize them in order to compare with a
+			// ConfigResource.
+			switch target := targetAddr.(type) {
 			case addrs.AbsResourceInstance:
-				targetAddr = instance.ContainingResource().Config()
-			case addrs.ModuleInstance:
-				targetAddr = instance.Module()
```

Comment:
I was curious about why the module case was removed here, and tried to check if targeting a module instance would work. It doesn't seem to (unless my test is broken), and I couldn't find a way to make it work, but I'm very unfamiliar with this code so that's not surprising.

Should `terraform apply -target=module.mod[1]` work? Does it already and I'm just mistaken?

Reply:
Thanks @alisdair! It appears I was thinking about this block in terms of only resource addresses (probably because of the case statement), and not the module itself. Let me add some tests in here too, since that was obviously not covered.